### PR TITLE
[WFCORE-6384] When shuting down to perform an installation CLI is dis…

### PR DESCRIPTION
**Pull request by yersan**
…connected from DC

Jira issue: https://issues.redhat.com/browse/WFCORE-6384

Instead of comparing Paths to understand whether we are using a local CLI client instance when performing an installation, now we use a CLI client marker file. The content of this client marker file is returned by the `shutdown --perform-installation` management operation. If the content doesn't match, we are not using a local CLI instance. Otherwise, we use the most restrictive approach assuming assuming the CLI is local.

### CODE REVIEW

File: host-controller/src/main/java/org/jboss/as/host/controller/operations/HostShutdownHandler.java
Comment (jfdenise) [inline host-controller/src/main/java/org/jboss/as/host/controller/operations/HostShutdownHandler.java:137]: We should test that the Files.exist

File: cli/src/main/java/org/jboss/as/cli/handlers/ShutdownHandler.java
Comment (jfdenise) [inline cli/src/main/java/org/jboss/as/cli/handlers/ShutdownHandler.java:187]: Not sure why we need to check for the presence of JBOSS_HOME env variable?

File: cli/src/main/java/org/jboss/as/cli/CommandLineMain.java
Comment (jfdenise) [inline cli/src/main/java/org/jboss/as/cli/CommandLineMain.java:183]: Could we have a warning displayed. I imagine the case where part of the JBOSS_HOME sub directories would be readonly.

File: host-controller/src/main/java/org/jboss/as/host/controller/operations/HostShutdownHandler.java
Comment (yersan) [reply host-controller/src/main/java/org/jboss/as/host/controller/operations/HostShutdownHandler.java:137]: I think we can do it without checking the existence since we can build Path objects for unknown files since Path doesn't access to the filesystem, do you agree?

File: cli/src/main/java/org/jboss/as/cli/handlers/ShutdownHandler.java
Comment (yersan) [reply cli/src/main/java/org/jboss/as/cli/handlers/ShutdownHandler.java:187]: Right, it was left over from a previous approach, we can remove the JBOSS_HOME check here

File: cli/src/main/java/org/jboss/as/cli/CommandLineMain.java
Comment (yersan) [reply cli/src/main/java/org/jboss/as/cli/CommandLineMain.java:183]: Yes, this is explicitly ignored because of the read-only use case (or any other exception).

We could log a WARN (we should move the method invocation after the one that setups the logging) but I considered it as a corner case that could go silently; if that occurs the logic will assume we are using a local CLI instance and it will get closed when the update is in place just because we won't be able to detect whether it is local.

We can add a WARN, but, shouldn't it be just noise since it is a very particular case only related to the shutdown when performing an installation? WDYT?

Comment (darranl) [reply cli/src/main/java/org/jboss/as/cli/CommandLineMain.java:183]: Or at the very least is any logging possible?  Catch blocks that drop the Throwable somehow end up being the place you need more information from when remote debugging an issue.

File: cli/src/main/java/org/jboss/as/cli/CommandLineMain.java
Comment (darranl) [inline cli/src/main/java/org/jboss/as/cli/CommandLineMain.java:168]: I am not sure we need the SecurityManager wrappers here, in this case as this is the entry point there should not be other protection domains on the call stack.

File: cli/src/main/java/org/jboss/as/cli/CommandLineMain.java
Comment (yersan) [reply cli/src/main/java/org/jboss/as/cli/CommandLineMain.java:171]: The RFE contemplates this situation where we have stated that is the customer the one that has to ensure no other processes are launched when they are updating the server.

In this case, the marker value returned by the customer won't match the current CLI instance, so it will assume it is not a local instance and won't be automatically closed.

The user will end up in the same situation as the current Patch tool, where an update/patch can update the underlying jboss-modules.jar. Since the client CLI session is not automatically closed, the user will be using an obsolete CLI process launched from a patched jboss-modules.jar. Probably he will get errors using the CLI and he will be forced to close it and open it again.

File: cli/src/main/java/org/jboss/as/cli/CommandLineMain.java
Comment (darranl) [reply cli/src/main/java/org/jboss/as/cli/CommandLineMain.java:168]: I am suggesting neither, just access the property directly.

Comment (darranl) [reply cli/src/main/java/org/jboss/as/cli/CommandLineMain.java:168]: The security manager being possibly enabled is never a reason to add a doPrivileged.

When you add a doPrivileged call what you are effectively doing is dropping the protection domains of any callers currently on the call stack and setting up the permission check to be in the context of the protection domain of the class that is calling the method with the permissions check.

As this is a main method / entry point - what callers are being dropped from the call stack and what measures have been put in place to verify they can safely call this method.

The way these properties are being used the safety is probably fine but what callers are being dropped from the call stack?

Comment (darranl) [reply cli/src/main/java/org/jboss/as/cli/CommandLineMain.java]: > I just wanted to avoid the need to require specific permission to load an environment variable that is indeed configured by the launch script. We know that this is safe to do because it is the launch script the one that invokes this Main class.

This is not the purpose of a doPrivileged call, it is not possible to prevent the permission check.  The ProtectionDomain for the CommandLineMain class will still need the relevant permissions to be granted.

When a permission check is performed it loops through the PermissionDomains of all classes on the call stack, and each and every one of those must be granted that permission or the check will fail.  What doPrivileged does is it says ignore the classes that called me but it does not get away from the "me" class being required the permission.

If this class is always intended to be called as the main entry point to the program it does not need any doPrivileged calls as it will always be the starting point of the call stack so there will be no protection domains to drop.

On the other hand if this is expected that other classes from other protection domains will call this class then this class should only be using doPrivileged if those are trusted called / it has been designed so the doPrivileged calls can not be abused.

TBH this pattern that creeps in where if there is going to be a permission check a doPrivileged is added is one of the reasons that adds to the justification of the removal of the security manager in Java.

Comment (yersan) [reply cli/src/main/java/org/jboss/as/cli/CommandLineMain.java]: I'm starting to understand the tricky details of the security manager when, for good reasons, it was going to be removed :)...never is late, even if it going to be an obsolete knowledge

**Comment by yersan** [comment]
@jfdenise / @darranl Could you take another look and approve if you think we can proceed as it is? Thanks!

**Comment by marekkopecky** [comment]
I confirm that the original issue described in https://issues.redhat.com/browse/WFCORE-6384 is fixed by this MR! Thank you!

File: cli/src/main/java/org/jboss/as/cli/CommandLineMain.java
Comment (jfdenise) [inline cli/src/main/java/org/jboss/as/cli/CommandLineMain.java:168]: @yersan when CLI is started as a fat jar (bin/client/jboss-cli-client.jar), we don't have this env set. I am wandering if we should attempt to find the jar location (class.getProtectionDomain().getCodeSource().getLocation()) and determinate if that is inside an installation?

Comment (yersan) [reply cli/src/main/java/org/jboss/as/cli/CommandLineMain.java:168]: @jfdenise We should be fine there since Installation Manager high level commands are not available for the CLI if the CLI is not running on a modular environment.

File: cli/src/main/java/org/jboss/as/cli/handlers/ShutdownHandler.java
Comment (jfdenise) [inline cli/src/main/java/org/jboss/as/cli/handlers/ShutdownHandler.java:205]: Shouldn't you return after this call?

Comment (yersan) [reply cli/src/main/java/org/jboss/as/cli/handlers/ShutdownHandler.java:205]: I've never found an issue, but yes, looks correct returning here.

**Review by darranl — APPROVED** [review]
Probably still needs review from @jfdenise but no objections from me.

**Review by jfdenise — APPROVED** [review]

File: cli/src/main/java/org/jboss/as/cli/CommandLineMain.java
Comment (jamezp) [inline cli/src/main/java/org/jboss/as/cli/CommandLineMain.java:47]: We're going to have to re-think this part and not use a static logger. This will be created before the `configureLogMangaer()` is invoked and could create issues.

Comment (jamezp) [reply cli/src/main/java/org/jboss/as/cli/CommandLineMain.java:47]: https://github.com/wildfly/wildfly-core/pull/5550

**Pull request merged**
